gateway/src/main.rs

```diff
@@ -237,46 +254,93 @@ async fn init(db: SqlitePool, args: InitArgs) -> io::Result<()> {
-
-            let identifier = format!("*.{public}");
-
-            // Use ::Dns01 challenge because that's the only supported
-            // challenge type for wildcard domains
-            let (chain, private_key) = acme
-                .create_certificate(&identifier, ChallengeType::Dns01, creds)
-                .await
-                .unwrap();
-
-            let mut buf = Vec::new();
-            buf.extend(chain.as_bytes());
-            buf.extend(private_key.as_bytes());
+async fn create_gateway_certificate<P: AsRef<Path>>(
+    fs: P,
+    public: FQDN,
+    acme: AcmeClient,
+    resolver: Arc<GatewayCertResolver>,
+) -> ChainAndPrivateKey {
+    let creds_path = fs.as_ref().join("acme.json");
+    let identifier = format!("*.{public}");
+    if !creds_path.exists() {
+        panic!(
+            "no ACME credentials found at {}, cannot continue with certificate creation",
+            creds_path.display()
+        );
+    }
+    let creds = std::fs::File::open(creds_path).expect("Invalid credentials path");
+    let creds: AccountCredentials =
+        serde_json::from_reader(&creds).expect("Can not parse admin credentials from path");
+    let account = Account::from_credentials(creds)
+        .map_err(|error| {
+            error!(
+                error = &error as &dyn std::error::Error,
+                "failed to convert acme credentials into account"
+            );
+            AcmeClientError::AccountCreation
+        })
+        .unwrap();
 
-            let certs = ChainAndPrivateKey::parse_pem(Cursor::new(buf)).unwrap();
+    // Use ::Dns01 challenge because that's the only supported
+    // challenge type for wildcard domains
+    let (chain, private_key) = acme
+        .create_certificate(&identifier, ChallengeType::Dns01, &account)
+        .await
+        .unwrap();
 
-            certs.clone().save_pem(&tls_path).unwrap();
+    let mut buf = Vec::new();
+    buf.extend(chain.as_bytes());
+    buf.extend(private_key.as_bytes());
+    let certs = ChainAndPrivateKey::parse_pem(Cursor::new(buf)).unwrap();
+    resolver.serve_default_der(certs.clone()).await.unwrap();
+    certs
+}
 
-            certs
+async fn automate_renewal<P: AsRef<Path> + Clone>(
+    fs: P,
+    public: FQDN,
+    acme: AcmeClient,
+    resolver: Arc<GatewayCertResolver>,
+) {
+    loop {
+        tokio::time::sleep(Duration::from_secs(240)).await;
+        let certs =
+            request_gateway_certs(fs.clone(), public.clone(), acme.clone(), resolver.clone()).await;
+        // Safe because a 'ChainAndPrivateKey' is built from a PEM.
+        let chain_and_pk = certs.into_pem().unwrap();
+        let (_, x509_cert) = parse_x509_certificate(chain_and_pk.as_bytes()).unwrap();
+        let diff = x509_cert.validity().not_after.sub(ASN1Time::now()).unwrap();
+        if diff.whole_days() <= 30 {
+            let tls_path = fs.as_ref().join("ssl.pem");
+            let certs = create_gateway_certificate(
```

Comment:
The strategy LGTM, but the only thing I'm concerned with is the challenge itself. Since this eventually calls `AcmeClient::create_certificate` with a challenge which is DNS-based, if the challenge key is different from the previous one this time round, then we'll need manual intervention and update our TXT records for `shuttleapp.rs` at the same time. Or give this service more AWS permissions so that it can update them on its own.

If we decide to go manual intervention (which is definitely easier), then better keep this behind an admin CLI subcommand that we fire ourselves. We can then "automate" it by having a simple (but separate) cron job that fires that admin subcommand once in a while, all we need this job to know about is what to update the TXT record to.

Reply:
Hah, you're right. The DNS challenge is a beast to secure [1]. This might need some dev work with AWS services, but I think before that we can start with introducing the support for requesting the gateway certificate through an admin command, on-demand, and leave the DNS challenge completion (required for certificate renewal) to be handled in a separate PR. What do you think?

[1] https://www.eff.org/deeplinks/2018/02/technical-deep-dive-securing-automation-acme-dns-challenge-validation

Reply:
I've implemented admin support to request for gateway certificate renewal (with DNS TXT record insertion) out of band in terms of API and cmdline. Let me know in case there is something else to address.